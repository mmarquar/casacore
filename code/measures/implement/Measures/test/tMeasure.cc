//# tMeasure.cc: This program test Measure functions
//# Copyright (C) 1995,1996
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This program is free software; you can redistribute it and/or modify it
//# under the terms of the GNU General Public License as published by the Free
//# Software Foundation; either version 2 of the License, or (at your option)
//# any later version.
//#
//# This program is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
//# more details.
//#
//# You should have received a copy of the GNU General Public License along
//# with this program; if not, write to the Free Software Foundation, Inc.,
//# 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

//# Includes
#include <aips/aips.h>
#include <aips/Exceptions/Error.h>
#include <aips/Measures.h>
#include <aips/Measures/Aberration.h>
#include <aips/Measures/MEpoch.h>
#include <aips/Measures/MDirection.h>
#include <aips/Measures/MPosition.h>
#include <aips/Measures/MRadialVelocity.h>
#include <aips/Measures/MVAngle.h>
#include <aips/Measures/MVTime.h>

main()
{
    try {
	cout << "Test measure class MVAngle" << endl;
	cout << "--------------------------------------" << endl;

	MVAngle mva1 = Quantity(190,"deg") + Quantity(59,"'") +
	    Quantity(59.95,"\"");
	cout << "Degrees:  " << mva1.degree() << endl;
	cout << "Radians:  " << mva1.radian() << endl;
	cout << "Fraction: " << mva1.circle() << endl;
	cout << "Degrees:  " << mva1().degree() << endl;
	cout << "Radians:  " << mva1().radian() << endl;
	cout << "Fraction: " << mva1().circle() << endl;
	cout << "Degrees:  " << mva1(+2).degree() << endl;
	cout << "Degrees:  " << mva1(-2).degree() << endl;
	cout << "Fraction: " << mva1(0).circle() << endl;
	cout << "Prec =2=: " << mva1.string(2) << endl;
	cout << "Prec =4=: " << mva1.string(4) << endl;
	cout << "Prec =6=: " << mva1.string(6) << endl;
	cout << "Prec =8=: " << mva1.string(8) << endl;
	cout << "Clean=2=: " << mva1.string(MVAngle::ANGLE_CLEAN, 2) << endl;
	cout << "Time = =: " << mva1.string(MVAngle::TIME) << endl;
	cout << "CNO_H= =: " << mva1.string(MVAngle::TIME_CLEAN_NO_H) << endl;

	MVTime mtim = 44362.6;
	cout << "Days:       " << mtim.day() << endl;    
	cout << "Hours:      " << mtim.hour() << endl;    
	cout << "Seconds:    " << mtim.second() << endl;    
	cout << "Minutes:    " << mtim.minute() << endl;    
	cout << "Prec =6=:   " << mtim.string(6) << endl;
	cout << "Day:        " << mtim.string((MVTime::formatTypes)
					      (MVTime::TIME | MVTime::DAY),
					      6) << endl;
	cout << "YMD:        " << mtim.string(MVTime::YMD, 6) << endl;
	cout << "DMY:        " << mtim.string(MVTime::DMY, 6) << endl;
	cout << "Day+DMY:    " << mtim.string((MVTime::formatTypes)
					      (MVTime::DMY | MVTime::DAY),
					      6) << endl;
	cout << "Day+YMD:    " << mtim.string((MVTime::formatTypes)
					      (MVTime::YMD | MVTime::DAY),
					      6) << endl;
	cout << "Weekday:    " << mtim.weekday() << endl;
	cout << "yyyy mm dd: " << mtim.year() << " " <<
	  mtim.month() << " " <<
	  mtim.monthday() << endl;
	cout << "yyyymmdd:   " << mtim.ymd() << endl;
	cout << "Yearday:    " << mtim.yearday() << endl;
	cout << "Yearweek:   " << mtim.yearweek() << endl;
	cout << "Week 960101:" << MVTime(1996,1,1).yearweek() << endl;
	cout << "Read it back:" << endl;
	mtim = MVTime(1980,5,3.6);
	cout << "Prec =6=:   " << mtim.string(6) << endl;
	cout << "Day:        " << mtim.string((MVTime::formatTypes)
					      (MVTime::TIME | MVTime::DAY),
					      6) << endl;
	cout << "YMD:        " << mtim.string(MVTime::YMD, 6) << endl;
	cout << "DMY:        " << mtim.string(MVTime::DMY, 6) << endl;
	cout << "Day+DMY:    " << mtim.string((MVTime::formatTypes)
					      (MVTime::DMY | MVTime::DAY),
					      6) << endl;
	cout << "Day+YMD:    " << mtim.string((MVTime::formatTypes)
					      (MVTime::YMD | MVTime::DAY),
					      6) << endl;
	cout << "Day (NO_T): " << mtim.string((MVTime::formatTypes)
					      (MVTime::TIME | MVTime::DAY |
					       MVTime::NO_TIME),
					      6) << endl;
	cout << "Day+YMD:    " << mtim.string((MVTime::formatTypes)
					      (MVTime::YMD | MVTime::DAY |
					       MVTime::NO_TIME),
					      6) << endl;
	cout << "Day+YMD:    " << mtim.string((MVTime::YMD | MVTime::DAY |
					       MVTime::NO_TIME),
					      6) << endl;

	cout << "Test measure class ..." << endl;
	cout << "--------------------------------------" << endl;

	MEpoch tbm(Quantity(MeasData::MJDB1950,"d"),MEpoch::Ref());
	cout << "Epoch B1950: " << tbm << endl;

	MEpoch::Ref tmref(MEpoch::TAI);
	MEpoch tm(Quantity(MeasData::MJD2000,"d"), tmref);

	cout << "Epoch 2000: " << tm << endl;
	cout << "Epoch reference: " << tmref << endl;

	MEpoch::Ref tbmref(MEpoch::TAI,tbm);
	MEpoch tm2(Quantity(MeasData::MJD2000-MeasData::MJDB1950,"d"),
		   tbmref);

	cout << "Epoch 2000 ref B1950: " << tm2 << endl;
	cout << "Epoch reference: " << tbmref << endl;

	cout << "Test measure conversion ..." << endl;
	cout << "--------------------------------------" << endl;
	MEpoch::Convert tconv(tm2,tmref);
	cout << "Converted " << tm2 << endl <<
	    " to " << tmref << endl <<
		" as " << tconv() << endl;
	MEpoch::Ref turef(MEpoch::UTC);
	MEpoch::Convert tconv2(tm2,turef);
	cout << "Converted " << tm2 << endl <<
	    " to " << turef << endl <<
		" as " << tconv2() << endl;

	MEpoch::Ref tgsref(MEpoch::UT1);
	MEpoch tm3(Quantity(50082.0,"d"),
		   tgsref);
	MEpoch tm4(Quantity(50082.72315521759259259259,"d"),
		   tgsref);
	MEpoch tm5(Quantity(44238.0,"d"),
		   tgsref);
	MEpoch::Ref tlsref(MEpoch::GMST1);
	MEpoch::Convert tconv3(tm3,tlsref);
	cout << "Converted " << tm3 << endl <<
	    " to " << tlsref << endl <<
		" as " << tconv3() << endl;
	cout << "Converted " << tm4 << endl <<
	    " to " << tlsref << endl <<
		" as " << tconv3(50082.72315521759259259259) << endl;
	cout << "Converted " << tm5 << endl <<
	    " to " << tlsref << endl <<
		" as " << tconv3(44238.0) << endl;
	MEpoch tm6(tconv3(44238.0));
	MEpoch::Convert tconv4(tm6,tgsref);
	cout << "Converted back " << tm6 << endl <<
	    " to " << tgsref << endl <<
		" as " << tconv4() << endl;

	MEpoch::Ref tasref(MEpoch::GAST);
	MEpoch::Convert tconv5(tm3,tasref);
	cout << "Converted " << tm3 << endl <<
	    " to " << tasref << endl <<
		" as " << tconv5() << endl;
	MEpoch tm7(tconv5());
	MEpoch::Convert tconv6(tm7,tgsref);
	cout << "Converted back " << tm7 << endl <<
	    " to " << tgsref << endl <<
		" as " << tconv6() << endl;

    {	
	MeasFrame b1900(MEpoch(Quantity(MeasData::MJDB1900,"d")));
	MDirection lsr1900(Quantity(270,"deg"),
			   Quantity(30,"deg"),
			   MDirection::Ref(MDirection::BMEAN,
					   b1900));
	cout << "LSR (B1900): " << 
	    lsr1900.getValue().getAngle("deg") << endl;
	cout << "LSR (B1950): " << 
	    MDirection::Convert(lsr1900, MDirection::B1950)()
		.getValue().getAngle("deg") << endl;
	cout << "LSR (J2000): " << 
	    MDirection::Convert(lsr1900, MDirection::J2000)()
		.getValue().getAngle("deg") << endl;
	cout << "LSR (B1900): " << 
	    lsr1900.getValue() << endl;
	cout << "LSR (B1950): " << 
	    MDirection::Convert(lsr1900, MDirection::B1950)()
		.getValue() << endl;
	cout << "LSR (J2000): " << 
	    MDirection::Convert(lsr1900, MDirection::J2000)()
		.getValue() << endl;
	cout << "LSR frame: " <<
	    MeasFrame(lsr1900) << endl;
    }

    {
	MDirection dirJ2000(Quantity(0,"deg"),
			    Quantity(30,"deg"),
			    MDirection::J2000);
	MEpoch ep(Quantity(50083.,"d"));
	MeasFrame frame(dirJ2000, ep);
	MRadialVelocity rvBary(Quantity(1000.,"km/s"),
			       MRadialVelocity::Ref
			       (MRadialVelocity::BARY,
				frame));
	cout << "Radial velocity (BARY): " << rvBary << endl <<
	    "                 (GEO): " <<
		MRadialVelocity::Convert(rvBary,
					 MRadialVelocity::GEO)()
					     .getValue() << endl;
	MRadialVelocity rvGeo = MRadialVelocity::Convert
	    (rvBary,
	     MRadialVelocity::GEO)();
	rvGeo.set(MRadialVelocity::Ref(MRadialVelocity::GEO, frame));
	cout << "and back: " <<
	    MRadialVelocity::Convert(rvGeo,
				     MRadialVelocity::BARY)()
					 .getValue() << endl;
	MPosition obs(Quantity(0,"m"),
		      Quantity(-289375.79,'"'), Quantity(50,"deg"));
	frame.set(obs);
	rvGeo.set(MVRadialVelocity(0.0));
	cout << "and 0 (GEO) to TOPO: " <<
	    MRadialVelocity::Convert(rvGeo,
				     MRadialVelocity::Ref
				     (MRadialVelocity::TOPO,
				      frame))()
					 .getValue() << endl;
	MRadialVelocity rvTopo = MRadialVelocity::Convert(rvGeo,
							  MRadialVelocity::Ref
							  (MRadialVelocity::TOPO,
							   frame))();
	cout << "and back: " <<
	    MRadialVelocity::Convert(rvTopo,
				     MRadialVelocity::GEO)()
					 .getValue() << endl;
	rvBary.set(MVRadialVelocity(0.0));
	cout << "and 0 (BARY) to LSR: " <<
	    MRadialVelocity::Convert(rvBary,
				     MRadialVelocity::Ref
				     (MRadialVelocity::LSR,
				      frame))()
					 .getValue() << endl;
	MRadialVelocity rvLSR = MRadialVelocity::Convert
	    (rvBary,
	     MRadialVelocity::Ref
	     (MRadialVelocity::LSR,
	      frame))();
	cout << "and back: " <<
	    MRadialVelocity::Convert(rvLSR,
				     MRadialVelocity::BARY)()
					 .getValue() << endl;

    }

    {
	cout << "Test real radial velocity" << endl <<
	    "-----------------------------------------" <<endl;
	MDirection coord(Quantity(185.425833,"deg"),
			 Quantity(31.799167,"deg"),
			 MDirection::B1950);
	MEpoch epo(Quantity(50217.625,"d"));
	MPosition pos(Quantity(10,"m"),
		      Quantity(6.60417,"deg"),
		      Quantity(52.8,"deg"),
		      MPosition::WGS84);
	MeasFrame frame(coord,epo,pos);
	cout << frame << endl;
	cout << MDirection::Convert(coord,
				    MDirection::Ref(MDirection::APP,
						    frame))()
							.getValue()
							    .getAngle("deg") <<
								endl;
	cout << MDirection::Convert(coord,
				    MDirection::Ref(MDirection::APP,
						    frame))()
							.getValue() <<
								endl;
	cout << MEpoch::Convert(epo,
				MEpoch::Ref(MEpoch::LAST,
					    frame))().getValue() 
						<< endl;
	MRadialVelocity rv(Quantity(1253,"km/s"),
			   MRadialVelocity::Ref(MRadialVelocity::LSR,
						frame));

	cout << MRadialVelocity::Convert(rv,
					 MRadialVelocity::Ref
					 (MRadialVelocity::TOPO,
					  frame))()
					      .getValue() -
						  rv.getValue() <<
						      endl;
 	epo = MEpoch(Quantity(50217.625,"d"));
	frame.set(epo);
	cout << MDirection::Convert(coord,
				    MDirection::Ref(MDirection::APP,
						    frame))()
							.getValue() <<
								endl;
	rv.set(MRadialVelocity::Ref(MRadialVelocity::BARY,
				    frame));
	Aberration ab2(Aberration::B1950);
	cout << "Aberration: " <<
	    ab2(epo.getValue().get()) << endl;
	cout << MRadialVelocity::Convert(rv,
					 MRadialVelocity::Ref
					 (MRadialVelocity::TOPO,
					  frame))()
					      .getValue() -
						  rv.getValue() <<
						      endl;
    }

    {
	MPosition obs(Quantity(0,"m"),
		      Quantity(-289375.79,'"'), Quantity(50,"deg"));
	MeasFrame frame(obs);
	MEpoch::Ref tasref(MEpoch::LAST, frame);
	MEpoch tm3(MVEpoch(50272.40590277778), MEpoch::UT1);
	MEpoch::Convert tconv5(tm3,tasref);
	MEpoch::Convert tconv51(tm3,MEpoch::GMST);
	MEpoch::Convert tconv52(tm3,MEpoch::GAST);
	cout << "Converted " << tm3 << endl <<
	    " to " << MEpoch::Ref(MEpoch::GMST) << endl <<
		" as " << tconv51() << endl;
	cout << "Converted " << tm3 << endl <<
	    " to " << MEpoch::Ref(MEpoch::GAST) << endl <<
		" as " << tconv52() << endl;
	cout << "Converted " << tm3 << endl <<
	    " to " << tasref << endl <<
		" as " << tconv5() << endl;
	MEpoch tm7(tconv5());
	MEpoch::Convert tconv6(tm7, MEpoch::UT1);
	cout << "Converted back " << tm7 << endl <<
	    " to " << MEpoch::Ref(MEpoch::UT1) << endl <<
		" as " << tconv6() << endl;

	MEpoch tm31(MVEpoch(50272.40590277778), MEpoch::UT1+MEpoch::RAZE);
	MEpoch::Ref tofref(MEpoch::LAST, frame, tm31);
	MEpoch::Convert tconv53(tm3,tofref);
	cout << "Converted " << tm3 << endl <<
	    " to " << tofref << endl <<
		" as " << tconv53() << endl;
	
    };	

    } catch (AipsError x) {
	cout << x.getMesg() << endl;
    } end_try;

    try {
	cout << "Test MeasFrame ..." << endl;
	cout << "--------------------------------------" << endl;

	MEpoch tbm(Quantity(MeasData::MJDB1950,"d"));
	cout << "Epoch B1950: " << tbm << endl;

        MeasFrame mftbm(tbm);
	cout << mftbm << endl;

    } catch (AipsError x) {
	cout << x.getMesg() << endl;
    } end_try;

    try {
	cout << "Test coordinate conversion ..." << endl;
	cout << "--------------------------------------" << endl;
	
	MDirection::Ref eqref(MDirection::B1950);
	MDirection::Ref galref(MDirection::GALACTIC);
	MDirection gpole(Quantity(192.25,"deg"),Quantity(27.4,"deg"),eqref); 
	MDirection eqpole(Quantity(192.25,"deg"),Quantity(27.4,"deg"),galref); 
	MDirection::Convert eqgal(gpole,galref);
	MDirection::Convert galeq(eqpole,eqref);
    

	cout << "Converted B1950 galactic pole " << gpole << endl << 
	    " to " << galref << endl <<
		" as " << eqgal().getAngle("deg") << endl;
	cout << "Converted B1950 galactic pole " << eqpole << endl << 
	    " to " << eqref << endl <<
		" as " << galeq().getAngle("deg") << endl;

	MEpoch app(Quantity(50083.0,"d"));
	MeasFrame appframe;
	MDirection::Ref appref(MDirection::APP, appframe);
	MDirection::Ref j2000ref(MDirection::J2000);
	MVDirection j2000vec(
			     -0.373798658,
			     -0.312643465,
			     -0.873228852);
	MDirection j2000(j2000vec);
	MDirection b1950(j2000vec, MDirection::B1950);
	MDirection::Convert j2000app(j2000,appref);
	appframe.set(app);
	MDirection appc(j2000app());
	MDirection::Convert appj2000(appc,j2000ref);
	MDirection::Convert b1950j2000(b1950, MDirection::J2000);
	MDirection appj(b1950j2000());
	MDirection::Convert j2000b1950(appj, MDirection::B1950);
	
	cout << "Converted J2000 coordinates " << j2000.getValue() <<
	    " to: " << j2000app().getValue() << endl;
	
	cout << "Converted back to J2000 coordinates " << appc.getValue() <<
	    " to: " << appj2000().getValue() << endl;
	
	cout << "Converted B1950 to J2000 coordinates " << b1950.getValue() <<
	    " to: " << b1950j2000().getValue() << endl;
	
	cout << "Converted J2000 to B1950 coordinates " << 
	    appj.getValue() <<
		" to: " << j2000b1950(appj).getValue() << endl;
	MPosition wsrt(Quantity(16,"m"), Quantity(6.60417,"deg"),
		       Quantity(52.91692,"deg"), MPosition::WGS84);
	MPosition::Convert wsrtitrf(wsrt, MPosition::ITRF);
	MPosition geod(wsrtitrf());

	cout << "Converted geodetic position: " << endl <<
	    wsrt.getValue().getLength() << ", " <<
		wsrt.getAngle("deg") << endl << "to: " <<
		    geod.getValue().getLength() << ", " <<
			geod.getAngle("deg") << endl;

	MPosition::Convert wsrtwgs(geod, MPosition::WGS84);
	wsrt = wsrtwgs();

	cout << "Converted geocentric position: " << endl <<
	    geod.getValue().getLength() << ", " <<
		geod.getAngle("deg") << endl << "to: " <<
		    wsrt.getValue().getLength() << ", " <<
			wsrt.getAngle("deg") << endl;

    {
	MEpoch tim(MVEpoch(50082), MEpoch::UT1);
	MeasFrame frame(wsrt, tim);
	MDirection appvec(MDirection::Convert(j2000vec, 
					      MDirection::Ref(MDirection::APP,
							      frame))());

	cout << "J2000 coordinates: " <<
	    j2000vec.getAngle("deg") << endl;
	cout << "Apparent coordinate" <<
	    appvec.getAngle("deg") << endl;
	Double d1, d2 , d3;
	frame.getLong(d1);
	frame.getLat(d2);
	frame.getLAST(d3);

	cout << "Longitude: " << MVAngle(d1).get("deg") << endl;
	cout << "Latitude:  " << MVAngle(d2).get("deg") << endl;
	cout << "LAST:      " << MVAngle(d3*C::circle).
	    string(MVAngle::TIME, 8) << endl;
	cout << "LAST:      " << MVAngle(d3*C::circle).
	    string(MVAngle::ANGLE, 8) << endl;
	cout << "LAST:      " << 
	    MEpoch::Convert(tim,
			    MEpoch::Ref(MEpoch::LAST, frame))() << endl;
	cout << "HA/DEC: " <<
	    MDirection::Convert(appvec, 
				MDirection::Ref(MDirection::HADEC,
						frame))()
						    .getAngle("deg") <<
							endl;
	MDirection had(MDirection::Convert(appvec, 
					   MDirection::Ref(MDirection::HADEC,
							   frame))());
	cout << "Back:   " <<
	    MDirection::Convert(had, 
				MDirection::Ref(MDirection::APP,
						frame))()
						    .getAngle("deg") <<
							endl;
	cout << "Az/El:  " <<
	    MDirection::Convert(had, 
				MDirection::Ref(MDirection::AZEL,
						frame))()
						    .getAngle("deg") <<
							endl;
	MDirection azel(MDirection::Convert(appvec, 
					   MDirection::Ref(MDirection::AZEL,
							   frame))());
	cout << "Back:   " <<
	    MDirection::Convert(azel, 
				MDirection::Ref(MDirection::HADEC,
						frame))()
						    .getAngle("deg") <<
							endl;
    };
    {
	MEpoch app(Quantity(47165.8,"d"));
	MeasFrame appframe(app);
	MDirection::Ref appref(MDirection::APP, appframe);
	MDirection::Ref b1950ref(MDirection::B1950);
	MDirection bappc(Quantity(85.4267,"deg"),
			 Quantity(49.8502,"deg"), appref);
	MDirection::Convert appb1950(bappc, b1950ref);
	MDirection b1950(appb1950());
	MDirection::Convert b1950app(b1950, appref);
	MVDirection mvpole;
	MDirection bappc1(Quantity(85.4267,"deg"),
			 Quantity(50.8502,"deg"), appref);
	
	cout << "Converted to B1950 coordinates " << bappc.getAngle("deg") <<
	    " to: " << appb1950().getAngle("deg") << endl;
	cout << "and back to: " <<
	    b1950app().getAngle("deg") << endl;
	MDirection apole(b1950app(mvpole));
	cout << "Rotation angle: " << 
	    bappc.getValue().positionAngle(apole.getValue(), "deg") << endl;
    };
	

    } catch (AipsError x) {
	cout << x.getMesg() << endl;
    } end_try;

    exit(0);
}
