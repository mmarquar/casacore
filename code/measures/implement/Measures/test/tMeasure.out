Test measure class MVAngle
--------------------------------------
Degrees:  191
Radians:  3.33358
Fraction: 0.530556
Degrees:  -169
Radians:  -2.94961
Fraction: -0.469444
Degrees:  911
Degrees:  -529
Fraction: 0.530556
Prec =2=: +191..
Prec =4=: +191.00.
Prec =6=: +191.00.00
Prec =8=: +190.59.59.95
Clean=2=:  191
Time = =: 12:43:59
CNO_H= =: 43:59
Days:       44362.6
Hours:      1.0647e+06
Seconds:    3.83293e+09
Minutes:    6.38821e+07
Prec =6=:   14:24:00
Day:        Sat-14:24:00
YMD:        1980/05/03/14:24:00
DMY:        03-May-1980/14:24:00
Day+DMY:    Sat-03-May-1980/14:24:00
Day+YMD:    Sat-1980/05/03/14:24:00
Weekday:    6
yyyy mm dd: 1980 5 3
yyyymmdd:   19800503
Yearday:    124
Yearweek:   18
Week 960101:1
Read it back:
Prec =6=:   14:24:00
Day:        Sat-14:24:00
YMD:        1980/05/03/14:24:00
DMY:        03-May-1980/14:24:00
Day+DMY:    Sat-03-May-1980/14:24:00
Day+YMD:    Sat-1980/05/03/14:24:00
Day (NO_T): Sat
Day+YMD:    Sat-1980/05/03
Day+YMD:    Sat-1980/05/03
Test measure class ...
--------------------------------------
Epoch B1950: Epoch: 33281::22:09:46.8619
Epoch 2000: Epoch: 51544::12:00:00.0000
Epoch reference: Reference for an Epoch with Type: TAI
Epoch 2000 ref B1950: Epoch: 18262::13:50:13.1381
Epoch reference: Reference for an Epoch with Type: TAI, Offset: Epoch: 33281::22:09:46.8619
Test measure conversion ...
--------------------------------------
Converted Epoch: 18262::13:50:13.1381
 to Reference for an Epoch with Type: TAI
 as Epoch: 51544::12:00:00.0000
Converted Epoch: 18262::13:50:13.1381
 to Reference for an Epoch with Type: UTC
 as Epoch: 51544::11:59:29.0000
Converted Epoch: 50082::00:00:00.0000
 to Reference for an Epoch with Type: GMST1
 as Epoch: 56791::06:35:48.3230
Converted Epoch: 50082::17:21:20.6108
 to Reference for an Epoch with Type: GMST1
 as Epoch: 56792::00:00:00.0000
Converted Epoch: 44238::00:00:00.0000
 to Reference for an Epoch with Type: GMST1
 as Epoch: 50931::06:35:18.7565
Converted back Epoch: 50931::06:35:18.7565
 to Reference for an Epoch with Type: UT1
 as Epoch: 44237::23:59:60.0000
Converted Epoch: 50082::00:00:00.0000
 to Reference for an Epoch with Type: GAST
 as Epoch: 56791::06:35:48.7519
Converted back Epoch: 56791::06:35:48.7519
 to Reference for an Epoch with Type: UT1
 as Epoch: 50081::23:59:60.0000
LSR (B1900): [-90, 30] deg
LSR (B1950): [-89.5206, 30.0012] deg
LSR (J2000): [-89.0405, 30.0047] deg
LSR (B1900): [-1.59086e-16, -0.866025, 0.5]
LSR (B1950): [0.00724658, -0.865985, 0.500018]
LSR (J2000): [0.0145021, -0.865863, 0.500071]
LSR frame: Frame: Direction: [-1.59086e-16, -0.866025, 0.5]
        (J2000 = [-89.0405, 30.0047] deg)
Radial velocity (BARY): Radialvelocity: 1e+06
                 (GEO): 1.02688e+06
and back: 999999
and 0 (GEO) to TOPO: 1.35506e-11
and back: 0
and 0 (BARY) to LSR: 3347.86
and back: 0
Test real radial velocity
-----------------------------------------
Frame: Epoch: 50217::15:00:00.0000 (TDB = 50217.6)
       Position: [6.00587, 0.695346, 7.9653]
        (Longitude = 0.115265 Latitude = 0.918296)
       Direction: [-0.846092, -0.0803642, 0.526943]
        (J2000 = [-173.951, 31.5223] deg)
[-173.993, 31.5433] deg
[-0.847565, -0.0891916, 0.523143]
56927::06:56:56.6228
16085.3
[-0.847565, -0.0891916, 0.523143]
Aberration: [7.87739e-05, -5.38236e-05, -2.33309e-05]
22015.2
Converted Epoch: 50272::09:44:30.0000
 to Reference for an Epoch with Type: GMST1
 as Epoch: 56982::04:50:59.8613
Converted Epoch: 50272::09:44:30.0000
 to Reference for an Epoch with Type: GAST
 as Epoch: 56982::04:51:00.1311
Converted Epoch: 50272::09:44:30.0000
 to Reference for an Epoch with Type: LAST,
Frame: Position: [1.07394e-07, -6.33753e-07, 7.66044e-07]
        (Longitude = -1.40293 Latitude = 0.872665)
 as Epoch: 56981::23:29:28.4118
Converted back Epoch: 56981::23:29:28.4118
 to Reference for an Epoch with Type: UT1
 as Epoch: 50272::09:44:30.0000
Converted Epoch: 50272::09:44:30.0000
 to Reference for an Epoch with Type: LAST, Offset: Epoch: 50272::09:44:30.0000,
Frame: Position: [1.07394e-07, -6.33753e-07, 7.66044e-07]
        (Longitude = -1.40293 Latitude = 0.872665)
 as Epoch: 0::23:29:28.4118
Test MeasFrame ...
--------------------------------------
Epoch B1950: Epoch: 33281::22:09:46.8619
Frame: Epoch: 33281::22:09:46.8619 (TDB = 33281.9)
Test coordinate conversion ...
--------------------------------------
Converted B1950 galactic pole Direction: [-0.867601, -0.188375, 0.4602]
 to Reference for an Direction with Type: GALACTIC
 as [-58.3295, 90] deg
Converted B1950 galactic pole Direction: [-0.867601, -0.188375, 0.4602]
 to Reference for an Direction with Type: B1950
 as [119.879, 29.4091] deg
Converted J2000 coordinates [-0.373799, -0.312643, -0.873229] to: [-0.374478, -0.31236, -0.873039]
Converted back to J2000 coordinates [-0.374478, -0.31236, -0.873039] to: [-0.373799, -0.312643, -0.873229]
Converted B1950 to J2000 coordinates [-0.373799, -0.312643, -0.873229] to: [-0.366031, -0.31678, -0.875027]
Converted J2000 to B1950 coordinates [-0.366031, -0.31678, -0.875027] to: [-0.373799, -0.312643, -0.873229]
Converted geodetic position: 
16 m, [6.60417, 52.9169] deg
to: 6.36458e+06 m, [6.60417, 52.7316] deg
Converted geocentric position: 
6.36458e+06 m, [6.60417, 52.7316] deg
to: 16 m, [6.60417, 52.9169] deg
J2000 coordinates: [-140.091, -60.836] deg
Apparent coordinate[-140.168, -60.8138] deg
Longitude: 6.60417 deg
Latitude:  52.7316 deg
LAST:      07:02:13.75
LAST:      +105.33.26.29
LAST:      Epoch: 56791::07:02:13.7527
HA/DEC: [-114.275, -60.8138] deg
Back:   [-140.168, -60.8139] deg
Az/El:  [129.705, -54.7033] deg
Back:   [-114.275, -60.8138] deg
Converted to B1950 coordinates [85.4267, 49.8502] deg to: [84.6813, 49.8286] deg
and back to: [85.4267, 49.8502] deg
Rotation angle: 0.335234 deg
